lib/evaluate/fold.cc

```diff
@@ -313,296 +316,383 @@ static std::optional<std::int64_t> GetInt64ArgOr(
   }
 }
 
+template<typename A, typename B>
+std::optional<std::vector<A>> GetIntegerVector(const B &x) {
+  static_assert(std::is_integral_v<A>);
+  if (const auto *someInteger{UnwrapExpr<Expr<SomeInteger>>(x)}) {
+    return std::visit(
+        [](const auto &typedExpr) -> std::optional<std::vector<A>> {
+          using T = ResultType<decltype(typedExpr)>;
+          if (const auto *constant{UnwrapConstantValue<T>(typedExpr)}) {
+            if (constant->Rank() == 1) {
+              std::vector<A> result;
+              for (const auto &value : constant->values()) {
+                result.push_back(static_cast<A>(value.ToInt64()));
+              }
+              return result;
+            }
+          }
+          return std::nullopt;
+        },
+        someInteger->u);
+  }
+  return std::nullopt;
+}
+
+// Transform an intrinsic function reference that contains user errors
+// into an intrinsic with the same characteristic but the "invalid" name.
+// This to prevent generating warnings over and over if the expression
+// gets re-folded.
+template<typename T> Expr<T> MakeInvalidIntrinsic(FunctionRef<T> &&funcRef) {
+  SpecificIntrinsic invalid{std::get<SpecificIntrinsic>(funcRef.proc().u)};
+  invalid.name = "invalid";
+  return Expr<T>{FunctionRef<T>{
+      ProcedureDesignator{std::move(invalid)}, std::move(funcRef.arguments())}};
+}
+
 template<typename T>
-ActualArguments &FoldArguments(
-    FoldingContext &context, FunctionRef<T> &funcRef) {
+Expr<T> Reshape(FoldingContext &context, FunctionRef<T> &&funcRef) {
+  auto args{funcRef.arguments()};
+  CHECK(args.size() == 4);
+  const auto *source{UnwrapConstantValue<T>(args[0])};
+  const auto *pad{UnwrapConstantValue<T>(args[2])};
+  std::optional<std::vector<ConstantSubscript>> shape{
+      GetIntegerVector<ConstantSubscript>(args[1])};
+  std::optional<std::vector<int>> order{GetIntegerVector<int>(args[3])};
+
+  if (!source || !shape || (args[2].has_value() && !pad) ||
+      (args[3].has_value() && !order)) {
+    return Expr<T>{std::move(funcRef)};  // Non-constant arguments
+  } else if (!IsValidShape(shape.value())) {
+    context.messages().Say("Invalid SHAPE in RESHAPE"_en_US);
+  } else {
+    int rank{GetRank(shape.value())};
+    std::size_t resultElements{TotalElementCount(shape.value())};
+    std::optional<std::vector<int>> dimOrder;
+    if (order.has_value()) {
+      dimOrder = ValidateDimensionOrder(rank, *order);
```

Comment:
Some instances of accessing `std::optional` variable values use `*` operator and some others use the `value()`. I feel we should try to keep it uniform. However, I do know that both kinds of accesses exist in the code base outside of this change.

Reply:
It is a fair point. The rational I try to follow is: `value()` will kill the program if the optional is empty `*` does not do this check, so I use `*` if the line just above is `if  (opt.has_value())` because the checks it obviously redundant, everywhere-else, I use `.value()` for safety. Exception to the exception of using `*` is when it would end up `&(*opt)` which I find disturbing...
And yes, both kinds of access already exists outside of this change.

Reply:
Please feel free to use `*` when it is obvious that the presence of the value does not have to be checked, as in this situation.  Using `.value()` on the line after checking `.has_value()` is not necessary.

Reply:
That rationale makes sense to me. Thanks.